### PR TITLE
helm: Replaced object-based extraArgs with array-based

### LINKED ISSUE #14944: Helm chart doesn't support multiple --api-rate-limit arguments
## Bug report

Hey guys,

Further to our [slack thread](https://cilium.slack.com/archives/C1MATJ5U5/p1613075990014900), I'm creating this issue to associate with the PR I'm about to submit.

The issue is that in order to adjust the default API rate limits, it's necessary to pass multiple `--api-rate-limit` arguments to cilium-agent. However, the helm chart implements `extraArgs` as an object, and so it's not possible to pass duplicate keys with different values. (duplicates are discarded).

**General Information**

- Cilium version (run `cilium version`) : 1.9.4
- Kernel version (run `uname -a`) : N/A
- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...) : Kubernetes with helm v3.5.2

**How to reproduce the issue**

```
❯ helm template . --set extraArgs.api-rate-limit=foo --set extraArgs.api-rate-limit=bar | grep rate-limit -C 3
      containers:
      - args:
        - --config-dir=/tmp/cilium/config-map
        - --api-rate-limit=bar
        command:
        - cilium-agent
        livenessProbe:
```

### LINKED ISSUE #15150: helm: `extraArgs` refactoring
Follow-up issue from discussion in https://github.com/cilium/cilium/pull/14974#issuecomment-786024908.

To do:

- [ ] Replace current top-level object-based `extraArgs` in `values.yaml` with array-based `extraArgs`, and edit `cilium-agent-daemonset.yaml` template accordingly
- [ ] Document Helm chart breaking change (release notes?)
- [ ] Implement and document identical array-based `extraArgs` for all sub-components

Implementing this will fix #14944.

### PR DESCRIPTION
**Pull request by sergeimonakhov**
Signed-off-by: Sergey Monakhov <monakhov@puzl.ee>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Hi, I refactored `extra-args`. 

Fixes: #15150

```release-noteaa
Replaced object-based extraArgs with array-based in Helm chart for all sub-components
```

### CODE REVIEW

**Comment by kkourt** [comment]
I remember that there was a discussion about this, so I found [this comment](https://github.com/cilium/cilium/pull/14974#issuecomment-786024908) that resulted in: https://github.com/cilium/cilium/issues/15150.

**Review by kkourt — APPROVED** [review]
Thanks!

File: install/kubernetes/cilium/values.yaml
Comment (nbusseneau) [inline install/kubernetes/cilium/values.yaml]: I think we should update this comment to:
```suggestion
  # -- Additional cilium-operator container arguments
```

**Review by nbusseneau — CHANGES_REQUESTED** [review]
Thanks, looks good! Since we are introducing a breaking change in Helm charts between 1.9 and this PR (1.10?), I would recommend documenting it directly in the commit. Could you add a mention for that in commit description? Something along the lines of:

> Breaking change: Helm releases previously deployed with additional arguments in object-based `values.extraArgs` will need to update these arguments to the new array-based format.

I suppose we may also need to add the breaking change to the release notes and/or upgrade guide (`Documentation/operations/upgrade.rst`), though I'm not sure about the process for either. @joestringer any input?

File: install/kubernetes/cilium/README.md
Comment (nbusseneau) [inline install/kubernetes/cilium/README.md]: Likewise:
```suggestion
| operator.extraArgs | list | `[]` | Additional cilium-operator container arguments |
```

**Comment by joestringer** [comment]
The "BPF checks" failure is really checking the commit description and is complaining about the format of the message. Typically in Cilium we will use a format like this for commit messages:

```
helm: Replace object-based extraArgs with array-based

Breaking change: Helm releases previously deployed with additional arguments in object-based "values.extraArgs", but this prevented providing the same parameter multiple times. Replace these arguments to the new array-based format.

Fixes: #15150
Signed-off-by: ...
```

You can also squash the commits together, I don't think they need to be separate commits. [This StackOverflow post goes into more detail](https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git).

> I suppose we may also need to add the breaking change to the release notes and/or upgrade guide (Documentation/operations/upgrade.rst), though I'm not sure about the process for either. @joestringer any input?

It'd be nice to do this, but the structure is not yet in place since we haven't prepared v1.10 upgrade docs yet. Maybe don't worry about it for now since this PR has `release-note/minor` which will show up in the release for folks to see. I'll take down a note to set up the v1.10 upgrade notes section and put a note in there for these changes.

**Review by nbusseneau — APPROVED** [review]
Thanks for the changes :) LGTM.

**Comment by Skymirrh** [comment]
@joestringer Not sure if we should mark this with `upgrade-impact` or not? On one hand, these options are really meant for development, on the other hand if anybody uses it and upgrades without looking at the release notes his Helm deployment will fail. But even with that said, it will fail at the chart validation stage, so it won't have an impact on the existing Helm deployment...

**Comment by joestringer** [comment]
@Skymirrh yeah I wouldn't worry too much in this case because the validation should pick it up super early. In general `upgrade-impact` label is useful to (1) communicate amongst developers that we should consider upgrade implications and figure out how to either address them or document them, or (2) (maybe) later on if there's an upgrade issue, search for all the issues with `upgrade-impact` and see if any of them might be relevant.

**Comment by Skymirrh** [comment]
test-me-please

**Comment by joestringer** [comment]
> Totally useless for this PR but let's get them green because why not.

Somehow I thought we were using extraArgs in CI, but apparently not. 

For what it's worth, if we *know* that a PR won't be validated by CI because it's too simple or we don't use the functionality (like here), we don't need to run the CI just to try to get some green ticks. It's good enough to post "No need to run CI as no tests use this functionality" (assuming you've validated this is true! :sweat_smile: ) . Then janitors/maintainers can recognize why the CI was not run and consider that before merging.

We did have a few of the jobs pass, gke-stable failed to provision and the `test-1.2.0-4.9` run appears to have flaked out on one Istio test run (I filed #15337 to follow up). Just waiting on kubernetes codeowners to review /cc @nathanjsweet

**Review by nathanjsweet — APPROVED** [review]

**Comment by joestringer** [comment]
@D1abloRUS did you have some example YAML snippets of how you configured this option? Or did we end up documenting this somewhere? It's a little hard to tell exactly what the correct syntax looks like just from reading the PR.

**Comment by nbusseneau** [comment]
> @D1abloRUS did you have some example YAML snippets of how you configured this option? Or did we end up documenting this somewhere? It's a little hard to tell exactly what the correct ConfigMap syntax looks like just from reading the PR.

Correct syntax is:
```yaml
extraArgs:
  - --foo=bar
  - --baz=qux
```
It's not documented anywhere AFAIK, but I personally felt it was OK since it's a developer-centric flag and the `{{- toYaml . | trim | nindent 8 }}` approach is kinda standard (we have it for a lot of list objects in the charts).

Do you think we should document it?

**Comment by michi-covalent** [comment]
for some settings we just have a sample configuration values commented out: https://github.com/cilium/cilium/blob/c95f6fda172cd93bf2848b17dcc8bfdbe5ed63f4/install/kubernetes/cilium/values.yaml#L177

this way at least you can kind of see how the helm value is supposed to look like.

**Pull request merged**
